File: js/src/lib/pages/LoginPage.ts
import { expect, Page } from '@playwright/test';
import { TestUser } from '../fixtures/testUsers';

export class LoginPage {
  constructor(private readonly page: Page) {}

  async goto(): Promise<void> {
    await this.page.goto('/login');
    await expect(this.page.getByRole('heading', { level: 1, name: /sign in|log in/i })).toBeVisible();
  }

  async fillForm(user: TestUser): Promise<void> {
    await this.page.getByLabel(/email/i).fill(user.email);
    await this.page.getByLabel(/password/i).fill(user.password);
  }

  async submit(): Promise<void> {
    await this.page.getByRole('button', { name: /sign in|log in/i }).click();
  }
}

File: py/src/playwright_adventures/pages/login_page.py
from playwright.async_api import Page, expect

from ..config import BASE_URL
from ..journeys.models import TestUser


class LoginPage:
    def __init__(self, page: Page) -> None:
        self.page = page

    async def goto(self) -> None:
        await self.page.goto(f"{BASE_URL}/login")
        await expect(self.page.get_by_role("heading", level=1, name=r"(?i)sign in|log in")).to_be_visible()

    async def fill_form(self, user: TestUser) -> None:
        await self.page.get_by_label(r"(?i)email").fill(user.email)
        await self.page.get_by_label(r"(?i)password").fill(user.password)

    async def submit(self) -> None:
        await self.page.get_by_role("button", name=r"(?i)sign in|log in").click()
